### PR TITLE
Make AQL UPSERT asynchronous

### PR DESCRIPTION
**Pull request by goedderz**
### Scope & Purpose

Make AQL UPSERT operations asynchronous.

- [X] :hankey: Bugfix
- [X] :pizza: New feature
- [X] :fire: Performance improvement

### Checklist

- [X] Tests
  - [X] **integration tests**
- [x] :book: CHANGELOG entry made

#### Related Information

Related PR: https://github.com/arangodb/arangodb/pull/21068

### CODE REVIEW

File: arangod/Aql/UpsertModifier.cpp [hunk 1]
```diff
@@ -362,11 +446,29 @@ size_t UpsertModifier::nrOfWritesIgnored() const { return nrOfErrors(); }
 size_t UpsertModifier::getBatchSize() const { return _batchSize; }
 
 bool UpsertModifier::hasResultOrException() const noexcept {
-  return resultState() == ModificationExecutorResultState::HaveResult;
+  return std::visit(overload{
```
Comment (jsteemann) [inline arangod/Aql/UpsertModifier.cpp:449]: I think this is a public method that is currently called only from the ModificationExecutor in an assertion. To be technically correct, I think this method must acquired the `_resultMutex` as well.
As the method is currently only used in a single assertion, we could alternatively remove it altogether.

File: arangod/Aql/UpsertModifier.cpp [hunk 2]
```diff
@@ -362,11 +446,29 @@ size_t UpsertModifier::nrOfWritesIgnored() const { return nrOfErrors(); }
...
+                        [](Waiting) { return false; },
+                        [](HaveResult) { return true; },
+                        [](std::exception_ptr const&) { return true; },
+                    },
+                    _results);
 }
 
 bool UpsertModifier::hasNeitherResultNorOperationPending() const noexcept {
-  return resultState() == ModificationExecutorResultState::NoResult;
+  return std::visit(overload{
```
Comment (jsteemann) [inline arangod/Aql/UpsertModifier.cpp:459]: I think this is a public method that is currently called only from the ModificationExecutor in an assertion. To be technically correct, I think this method must acquired the `_resultMutex` as well.
As the method is currently only used in a single assertion, we could alternatively remove it altogether.

File: arangod/Aql/UpsertModifier.cpp [hunk 3]
```diff
@@ -277,49 +277,133 @@ void UpsertModifier::accumulate(InputAqlItemRow& row) {
...
   auto toInsert = _insertAccumulator.closeAndGetContents();
   if (toInsert.isArray() && toInsert.length() > 0) {
-    _insertResults =
-        trx.insert(_infos._aqlCollection->name(), toInsert, _infos._options);
+    auto future = [&] {
+      auto guard = std::lock_guard(_trxMutex);
+      if (!_trxAlive) {
+        THROW_ARANGO_EXCEPTION(TRI_ERROR_QUERY_KILLED);
+      }
+      return trx.insertAsync(_infos._aqlCollection->name(), toInsert,
```
Comment (jsteemann) [inline arangod/Aql/UpsertModifier.cpp:382]: Is it intended that we hold `_trxMutex` here while executing the `insertAsync` call?
Same question for the following calls to `updateAsync` / `replaceAsync`.

File: arangod/Aql/UpsertModifier.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/Aql/UpsertModifier.cpp:449]: I leave the decision up to you. If no mutex lock is added here, I guess a code comment would still make sense so others aren't confused about the (intentionally) missing protection here.

File: arangod/Aql/UpsertModifier.cpp (hunk 3 above)
Comment (goedderz) [reply arangod/Aql/UpsertModifier.cpp:382]: Yes, this is intentional. This serializes the calls to `insertAsync` (etc.) with `stopAndClear()`, which in turn makes sure that the transaction methods aren't destroyed during the call.
Note that `co_await`ing the Future `insertAsync` (etc.) returns is done after releasing the mutex, so it's not held during network requests. The transaction Methods themselves have to make sure that the actually asynchronous part of the code handles a concurrent destroy of the Methods properly (which my previous PRs should ensure).

File: arangod/Aql/UpsertModifier.cpp (hunk 1 above)
Comment (goedderz) [reply arangod/Aql/UpsertModifier.cpp:452]: I've added comments and left it without the mutex.

**Review by jsteemann — APPROVED** [review]
LGTM

**Pull request merged**